### ISSUE TITLE
Parameter mismatch in EXLA computation

### ISSUE THREAD

**Issue by seanmor5** [elixir-nx/nx#1260]
I have a weird failing test case in Axon for loss scales that leads to this error:

```
  7) test mixed precision training integration trains model with batch norm with policy compute f16, params f16, output f16, scale :dynamic (Axon.IntegrationTest)
     test/axon/integration_test.exs:437
     ** (RuntimeError) Call parameter must match argument; got parameter 1 shape: u8[2], argument shape: pred[2].
     stacktrace:
       (exla 0.5.3) lib/exla/op.ex:855: EXLA.Op.unwrap!/1
       (exla 0.5.3) lib/exla/op.ex:109: EXLA.Op.get_shape/1
       (exla 0.5.3) lib/exla/defn.ex:1744: EXLA.Defn.op_type/1
       (exla 0.5.3) lib/exla/defn.ex:1748: EXLA.Defn.to_type/2
       (exla 0.5.3) lib/exla/defn.ex:1553: EXLA.Defn.to_aggregate/7
       (exla 0.5.3) lib/exla/defn.ex:599: EXLA.Defn.cached_recur_operator/4
       (exla 0.5.3) lib/exla/defn.ex:481: EXLA.Defn.recur_operator/3
       (elixir 1.14.2) lib/enum.ex:1780: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
       (exla 0.5.3) lib/exla/defn.ex:598: EXLA.Defn.cached_recur_operator/4
       (exla 0.5.3) lib/exla/defn.ex:481: EXLA.Defn.recur_operator/3
       (elixir 1.14.2) lib/enum.ex:1780: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
       (elixir 1.14.2) lib/enum.ex:1780: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
       (exla 0.5.3) lib/exla/defn.ex:598: EXLA.Defn.cached_recur_operator/4
       (exla 0.5.3) lib/exla/defn.ex:481: EXLA.Defn.recur_operator/3
       (elixir 1.14.2) lib/enum.ex:1780: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
       (elixir 1.14.2) lib/enum.ex:1780: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
       (exla 0.5.3) lib/exla/defn.ex:598: EXLA.Defn.cached_recur_operator/4
       (exla 0.5.3) lib/exla/defn.ex:481: EXLA.Defn.recur_operator/3
       (elixir 1.14.2) lib/enum.ex:1780: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
       (elixir 1.14.2) lib/enum.ex:1780: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
```

I think the culprit is:

```elixir
  defnp adjust_dynamic(grads, %{loss_scale: loss_scale, counter: counter}, opts \\ []) do
    opts = keyword!(opts, period: 2_000, factor: 2, min_loss_scale: 1)

    grads_are_finite =
      deep_reduce(grads, Nx.tensor(1), fn x, acc ->
        x
        |> is_finite()
        |> Nx.logical_and(acc)
      end)

    new_loss_scale =
      if grads_are_finite do
        if counter == opts[:period] - 1 do
          first_finite(loss_scale * opts[:factor], loss_scale)
        else
          loss_scale
        end
      else
        Nx.max(opts[:min_loss_scale], loss_scale / opts[:factor])
      end

    new_counter = Nx.remainder(counter + 1, opts[:period]) * grads_are_finite

    %{loss_scale: new_loss_scale, counter: new_counter}
  end
```

triggered by running `USE_EXLA=true mix test test/axon/integration_test.exs --include integration`

**Comment by josevalim** [comment]
I was able to isolate it to this test:

```

    defn logical_and(a, b, c) do
      Nx.logical_and(
        is_finite(c),
        Nx.logical_and(is_finite(a), Nx.logical_and(is_finite(b), 1))
      )
    end

    defnp is_finite(x), do: Nx.all(Nx.logical_not(Nx.is_infinity(x)))

    test "logical and and select" do
      assert_equal(logical_and(1, Nx.tensor(0.0) |> Nx.broadcast({2}), 2.0), Nx.u8(1))
      flunk("should have failed earlier")
    end
```

I will further reduce it and fix it later. :)

**Comment by josevalim** [comment]
Fixed in aadd6318ae4423c82e21d59624c0ecdea6689dbb.
